src/core/shared_memory_manager.cc

```diff
@@ -0,0 +1,244 @@
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "src/core/shared_memory_manager.h"
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <algorithm>
+#include <deque>
+#include <exception>
+#include <future>
+#include "src/core/constants.h"
+#include "src/core/logging.h"
+#include "src/core/server_status.h"
+
+namespace nvidia { namespace inferenceserver {
+
+namespace {
+
+Status
+OpenSharedMemoryRegion(const std::string& shm_key, int* shm_fd)
+{
+  // get shared memory region descriptor
+  *shm_fd = shm_open(shm_key.c_str(), O_RDONLY, S_IRUSR | S_IWUSR);
+  if (*shm_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to get shared memory descriptor");
+  }
+
+  return Status::Success;
+}
+
+Status
+GetSharedMemoryAddress(
+    const int shm_fd, const size_t offset, const size_t byte_size,
+    void** shm_addr)
+{
+  // map shared memory to process address space
+  *shm_addr = mmap(NULL, byte_size, PROT_WRITE, MAP_SHARED, shm_fd, offset);
+  if (*shm_addr == MAP_FAILED) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to process address space");
+  }
+
+  return Status::Success;
+}
+
+Status
+CloseSharedMemoryRegion(int shm_fd)
+{
+  int tmp = close(shm_fd);
+  if (tmp == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to close shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+MmapCleanup(void* shm_addr, size_t byte_size)
+{
+  int tmp_fd = munmap(shm_addr, byte_size);
+  if (tmp_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to munmap shared memory region");
+  }
+
+  return Status::Success;
+}
+
+}  // namespace
+
+Status
+SharedMemoryManager::RegisterSharedMemory(
+    const std::string& name, const std::string& shm_key, const size_t offset,
+    const size_t byte_size)
+{
+  LOG_VERBOSE(1) << "Register() shared memory region: '" << name << "'";
+
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(register_mu_);
+
+  // If key is already in shared_memory_map_ then return error saying already
+  // registered
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "The shared memory region '" + name + "' is already registered.");
+  }
+
+  // register (or re-register)
+  try {
+    void* tmp_addr;
+    int shm_fd = -1;
+
+    // don't re-open if shared memory is already open
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key)
+        shm_fd = itr->second->shm_fd_;
+      else {
+        RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
+      }
+    }
+
+    RETURN_IF_ERROR(
+        GetSharedMemoryAddress(shm_fd, offset, byte_size, &tmp_addr));
```

Comment:
How about MapSharedMemory? Can you use something more descriptive than tmp_addr. Perhaps mapped_addr.